src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -443,12 +550,38 @@ public bool Equals(TypeSymbolWithAnnotations other, TypeCompareKind comparison)
 
             if ((comparison & TypeCompareKind.IgnoreNullableModifiersForReferenceTypes) == 0)
             {
-                var thisIsNullable = IsNullable;
-                var otherIsNullable = other.IsNullable;
-                if (otherIsNullable != thisIsNullable)
+                var thisAnnotation = NullableAnnotation;
+                var otherAnnotation = other.NullableAnnotation;
+                if (otherAnnotation != thisAnnotation)
                 {
-                    if ((comparison & TypeCompareKind.UnknownNullableModifierMatchesAny) == 0 ||
-                        (thisIsNullable != null && otherIsNullable != null))
+                    if (thisAnnotation == NullableAnnotation.Unknown || otherAnnotation == NullableAnnotation.Unknown)
+                    {
+                        if ((comparison & TypeCompareKind.UnknownNullableModifierMatchesAny) == 0)
+                        {
+                            return false;
+                        }
+                    }
+                    else if ((comparison & TypeCompareKind.IgnoreInsignificantNullableModifiersDifference) == 0)
+                    {
+                        return false;
+                    }
+                    else if (thisAnnotation.IsAnyNullable())
+                    {
+                        if (!otherAnnotation.IsAnyNullable())
+                        {
+                            return false;
+                        }
+                    }
+                    else if (!otherAnnotation.IsAnyNullable())
+                    {
+                        Debug.Assert(thisAnnotation.IsAnyNotNullable());
+                        Debug.Assert(otherAnnotation.IsAnyNotNullable());
+                        if (TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter())
```

Comment:
>if (TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter()) [](start = 24, length = 62)

What if `other.TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter()` instead? Do we return `false` in that case?

Reply:
> What if other.TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter() instead? Do we return false in that case?

If we reach here types are the same, they are compared at the beginning of this function.

---
In reply to: [232119017](https://github.com/dotnet/roslyn/pull/30913#discussion_r232119017) [](ancestors = 232119017)